### ISSUE TITLE
Give support to an OGC API Tiles layer

### ISSUE THREAD

**Issue by doublebyte1** [qgis/QGIS#50296]
### Feature description

With the publication of part 1 of [OGC API - Tiles](http://docs.ogc.org/DRAFTS/20-057.html#rc_tiles_core), it would be great if QGIS would support adding layers published using this standard.

The idea would be to allow users to use the browser panel or the layer menu, to connect to a conformant server, and pull OGC API Tiles layers. OGC API Tiles supports vector, map and coverage tiles, but I think that it would be easier to start with vector tiles.

### Additional context
If the community thinks this useful, I could start working on a PR.

**Comment by jerstlouis** [comment]
@doublebyte1 This was already working when we tested in 2021 during the _Modular OGC API Workflows_ project using the GDAL [OGC API driver](https://gdal.org/drivers/raster/ogcapi.html) which is available in GDAL 3.2+. Maps, Coverages, Tiles (including Map tiles, Coverages tiles, Vector tiles) were all working well.

From the menu bar, you select _Layer_ > _Add Layer_ > _Vector_ ; keep _File_ radio button selected, and for _Vector dataset_ prefix the URL to either the landing page or an individual collection by `OGCAPI:`. This process should probably be made more intuitive / obvious and better documented.

Then you can select a preferred API, format, TileMatrixSet as per the opening options of the OGC API driver.

See (draft) demo video at https://www.dailymotion.com/video/x80xzex (password: _ModularWorkflows_) at ~18:30 timestamp.

It would be nice to experiment with this again at the [Web Mapping Code sprint next week](https://developer.ogc.org/sprints/19/) to make sure it still works as expected.

**Comment by tomkralidis** [comment]
Given the above, we should also consider overall OGC API workflow in QGIS (`Layer -> Add OGC API Layer`?).

**Comment by doublebyte1** [comment]
> 

I think that is a great idea. We could add a collection endpoint, and QGIS would give the option to render it in whichever OAPIs are available.

**Comment by doublebyte1** [comment]
> @doublebyte1 This was already working when we tested in 2021 during the _Modular OGC API Workflows_ project using the GDAL [OGC API driver](https://gdal.org/drivers/raster/ogcapi.html) which is available in GDAL 3.2+. Maps, Coverages, Tiles (including Map tiles, Coverages tiles, Vector tiles) were all working well. (NOTE: Even though the _ogcapi_ GDAL driver is classified as raster, it also supports vector tiles. There is a comment saying that it could perhaps eventually also be merged with the _[oapif](https://gdal.org/drivers/vector/oapif.html#vector-oapif)_ _OGC API - Features_ driver).
> 
> From the menu bar, you select _Layer_ > _Add Layer_ > _Vector_ (for vector tiles) or _Raster_ (for Maps / Coverage (Tiles)) ; keep _File_ radio button selected, and for _Vector dataset_ prefix the URL to either the landing page or an individual collection by `OGCAPI:`. This process should probably be made more intuitive / obvious and better documented.
> 
> Then you can select a preferred API, format, TileMatrixSet as per the opening options of the OGC API driver.
> 
> See (draft) demo video at https://www.dailymotion.com/video/x80xzex (password: _ModularWorkflows_) at ~18:30 timestamp.
> 
> It would be nice to experiment with this again at the [Web Mapping Code sprint next week](https://developer.ogc.org/sprints/19/) to make sure it still works as expected.
> 
> EDIT: Just tested, and this still works: `OGCAPI:https://maps.gnosis.earth/ogcapi/collections/blueMarble` Not as much luck with `OGCAPI:https://maps.gnosis.earth/ogcapi/collections/Daraa:AgricultureSrf` ...

@jerstlouis I suppose this is the matching (not 100% clear from the documentation of the driver):
* MAPS - OGC API Maps
* TILES - OGC API Vector tiles- could it also be other tiles (e.g.: map, raster)?
* COVERAGE - OGC API Coverages
* ITEMS - OGC API Features?

**Comment by jerstlouis** [comment]
@doublebyte1 Yes, correct, that is the mapping of the resources / access mechanisms to the OGC API Standards.

As you say, Tiles is not only for Vector Tiles, but also for Coverage tiles and Map Tiles.

Note that one of the first things that should be fixed is this issue about supporting 2DTMS 2.0 in GDAL:

https://github.com/OSGeo/gdal/issues/6882

which really should not be too difficult, so a good thing to tackle at this upcoming code sprint.
Happy to provide additional 2DTMS guidance to anyone working on this.

**Comment by doublebyte1** [comment]
@jerstlouis that sounds good. I have tested it and it works for:
* OGC API Features, OGC API Coverages
It did not work for OGC API Tiles (probably because of the issue you mentioned), but I could not get it to work with OGC API Maps, either.

**Comment by doublebyte1** [comment]
I created a new issue here: https://github.com/qgis/QGIS/issues/52674

**Comment by doublebyte1** [comment]
I have GDAL 3.6.2, released 2023/01/02 but it does not work. 

```
Unsupported Data Source: OGCAPI:https://maps.gnosis.earth/ogcapi/collections/blueMarbl|option:API=MAP|option:IMAGE_FORMAT=JPEG is not a supported raster data source 
GNOSIS Map Server (OGCAPI) - 400 Bad Request
Invalid data collection 
```

**Comment by agiudiceandrea** [comment]
@doublebyte1 I can confirm that it works using either QGIS 3.28.4 with GDAL 3.6.2 or QGIS 3.28.5 with GDAL 3.6.3.

**Comment by doublebyte1** [comment]
@agiudiceandrea Thank You! It seems to be a problem with the mac os package, which uses an older version of GDAL.

**Comment by doublebyte1** [comment]
@jerstlouis I have submitted this to address some issues: https://github.com/OSGeo/gdal/pull/7656
I'll have a look at https://github.com/OSGeo/gdal/issues/6882, once this is merged.

**Comment by doublebyte1** [comment]
> @doublebyte1 This was already working when we tested in 2021 during the _Modular OGC API Workflows_ project using the GDAL [OGC API driver](https://gdal.org/drivers/raster/ogcapi.html) which is available in GDAL 3.2+. Maps, Coverages, Tiles (including Map tiles, Coverages tiles, Vector tiles) were all working well. (NOTE: Even though the _ogcapi_ GDAL driver is classified as raster, it also supports vector tiles. There is a comment saying that it could perhaps eventually also be merged with the _[oapif](https://gdal.org/drivers/vector/oapif.html#vector-oapif)_ _OGC API - Features_ driver).
> 
> From the menu bar, you select _Layer_ > _Add Layer_ > _Vector_ (for vector tiles) or _Raster_ (for Maps / Coverage (Tiles)) ; keep _File_ radio button selected, and for _Vector dataset_ prefix the URL to either the landing page or an individual collection by `OGCAPI:`. This process should probably be made more intuitive / obvious and better documented.
> 
> Then you can select a preferred API, format, TileMatrixSet as per the opening options of the OGC API driver.
> 
> See (draft) demo video at https://www.dailymotion.com/video/x80xzex (password: _ModularWorkflows_) at ~18:30 timestamp.
> 
> It would be nice to experiment with this again at the [Web Mapping Code sprint next week](https://developer.ogc.org/sprints/19/) to make sure it still works as expected.
> 
> EDIT: Just tested, and this still works: `OGCAPI:https://maps.gnosis.earth/ogcapi/collections/blueMarble` Not as much luck with `OGCAPI:https://maps.gnosis.earth/ogcapi/collections/Daraa:AgricultureSrf` ...

This seems to be raster tiles, only? `OGCAPI:https://maps.gnosis.earth/ogcapi/collections/blueMarble`. 
For testing vector tiles I used this: `OGCAPI:https://maps.gnosis.earth/ogcapi/collections/Daraa`

**Comment by agiudiceandrea** [comment]
> Thank You! It seems to be a problem with the mac os package, which uses an older version of GDAL.

@doublebyte1, so, was your previous comment:

>I have GDAL 3.6.2, released 2023/01/02 but it does not work.
>Unsupported Data Source: >```OGCAPI:https://maps.gnosis.earth/ogcapi/collections/blueMarble|option:API=MAP|option:IMAGE_FORMAT=JPEG is not a supported raster data source API MAP requested, but not available```

incorrect about the GDAL version used?

**Comment by jerstlouis** [comment]
@doublebyte1 `blueMarble` provides raster tiles only (and the reason it was working when I tested that time was because it was using _OGC API - Coverages_).

`OGCAPI:https://maps.gnosis.earth/ogcapi/collections/Daraa:AgricultureSrf` provides vector tiles for a single layer, whereas 
`OGCAPI:https://maps.gnosis.earth/ogcapi/collections/Daraa` provides multi-layer vector tiles, including the _Agricultural Surfaces layer_ (both options should work).

**Comment by doublebyte1** [comment]
> > Thank You! It seems to be a problem with the mac os package, which uses an older version of GDAL.
> 
> @doublebyte1, so, was your previous comment:
> 
> > I have GDAL 3.6.2, released 2023/01/02 but it does not work.
> > Unsupported Data Source: >`OGCAPI:https://maps.gnosis.earth/ogcapi/collections/blueMarble|option:API=MAP|option:IMAGE_FORMAT=JPEG is not a supported raster data source API MAP requested, but not available`
> 
> incorrect about the GDAL version used?

@agiudiceandrea yes, that is correct. I was telling you the gdal version that I have in the system, which is different from the one QGIS uses.

**Comment by doublebyte1** [comment]
> @doublebyte1 `blueMarble` provides raster tiles only (and the reason it was working when I tested that time was because it was using _OGC API - Coverages_).
> 
> `OGCAPI:https://maps.gnosis.earth/ogcapi/collections/Daraa:AgricultureSrf` provides vector tiles for a single layer, whereas `OGCAPI:https://maps.gnosis.earth/ogcapi/collections/Daraa` provides multi-layer vector tiles, including the _Agricultural Surfaces layer_ (both options should work).

With the fix I was able to load raster tiles in QGIS, but not vector tiles. Maybe because of the issue you described here: https://github.com/OSGeo/gdal/pull/7656/files/955d044b188ee338f923000a3bf161530cecb7a7#r1181229171

![Screenshot from 2023-04-30 14-23-26](https://user-images.githubusercontent.com/1038897/235356673-6f5b91a4-fdf7-4c58-b5e8-bd8fc118386f.png)

**Comment by doublebyte1** [comment]
With this https://github.com/OSGeo/gdal/pull/7656 , we can read OGC API (raster) tiles in QGIS, at least from WorldCRS84Quad. I will close this issue for now.

**Comment by doublebyte1** [comment]
This appears to have stopped working again:

<img width="1440" alt="Screenshot 2023-12-11 at 20 18 21" src="https://github.com/qgis/QGIS/assets/1038897/27dd5368-3e0d-443c-9e18-3986a1c98068">

```
2023-12-11T20:18:44     CRITICAL    Unsupported Data Source : OGCAPI:https://maps.gnosis.earth/ogcapi/collections/blueMarble/tiles|option:API=TILES|option:TILEMATRIXSET=WorldCRS84Quad is not a supported raster data source
             Missing bbox
```

**Comment by jerstlouis** [comment]
@doublebyte1 The supported end-point for GDAL / QGIS are either the landing page (ending here at `/ogcapi`) or the collection (ending at `/blueMarble`)

Does it work if you remove the `/tiles` part? (which is the _list of map tilesets_).

**Comment by doublebyte1** [comment]
> @doublebyte1 The supported end-point for GDAL / QGIS are either the landing page (ending here at `/ogcapi`) or the collection (ending at `/blueMarble`)
> 
> Does it work if you remove the `/tiles` part? (which is the _list of map tilesets_).
> 
> (unless new support for _tileset lists_ end-points was added more recently to GDAL and I am not aware of it).

No, it doesn't. 

```
Unsupported Data Source : OGCAPI:https://maps.gnosis.earth/ogcapi/collections/blueMarble|option:API=TILES|option:TILEMATRIXSET=WorldCRS84Quad is not a supported raster data source
             API TILES requested, but not available
```

**Comment by doublebyte1** [comment]
When opening the landing page, it does detect the layers, but then it does not manage to open any.

<img width="1440" alt="Screenshot 2023-12-12 at 10 06 41" src="https://github.com/qgis/QGIS/assets/1038897/721d11bf-8c21-437c-bd3c-9fc0668d8093">

**Comment by kudlav** [comment]
Hi, I'm testing the compatibility of QGIS with OGC API Tiles to access tiled data and I'm not very successful.

### Loading by tileset description
_not supported_

Examples are from OpenLayers: [raster](https://openlayers.org/en/latest/examples/ogc-map-tiles.html), [vector](https://openlayers.org/en/latest/examples/ogc-vector-tiles.html).

- ❌ _Add raster_: https://maps.gnosis.earth/ogcapi/collections/blueMarble/map/tiles/WebMercatorQuad_
  - -> `Unsupported Data Source: OGCAPI:https://maps.gnosis.earth/ogcapi/collections/blueMarble/map/tiles/WebMercatorQuad is not a supported raster data source Missing bbox`
  - manually add some bbox: https://kudlac.tode.cz/shared/ogcapi-collections-blueMarble-map-tiles-WebMercatorQuad.json
  - -> `Unsupported Data Source: OGCAPI:https://kudlac.tode.cz/shared/ogcapi-collections-blueMarble-map-tiles-WebMercatorQuad.json is not a supported raster data source API AUTO requested, but not available`
- ❌ _Add vector_: https://maps.gnosis.earth/ogcapi/collections/NaturalEarth:cultural:ne_10m_admin_0_countries/tiles/WebMercatorQuad_
  - -> `Invalid Data Source: OGCAPI:https://maps.gnosis.earth/ogcapi/collections/NaturalEarth:cultural:ne_10m_admin_0_countries/tiles/WebMercatorQuad is not a valid or recognized data source`

### Loading from the landing page

- ❌ _Add Vector_: `Invalid Data Source: OGCAPI:https://maps.gnosis.earth/ogcapi is not a valid or recognized data source.`
- ❔  _Add Raster_ https://maps.gnosis.earth/ogcapi freezes the QGIS

### Loading from the collection description

- ❌ _Add Vector_: `Invalid Data Source: OGCAPI:https://maps.gnosis.earth/ogcapi/collections/Daraa is not a valid or recognized data source.`
- ❌ _Add Vector_: `Invalid Data Source: OGCAPI:https://maps.gnosis.earth/ogcapi/collections/NaturalEarth is not a valid or recognized data source.`
- Adding raster:
  - ✅  _Add Raster_ https://maps.gnosis.earth/ogcapi/collections/Daraa, reads successfully `Daraa:MilitarySrf/map/tiles/WorldCRS84Quad` tiles (png, jpg, tiff) 
  - ❌ _Add Raster_ https://maps.gnosis.earth/ogcapi/collections/blueMarble From the raster collection description QGIS can read only coverage TIFF, not the tiles or other formats like PNG or JPG. The following part of the JSON is actually read. If you remove the geotiff link from links of the original blueMarble json, it will stop working.

```json
{
  "dataType": "map",
  "crs": "http://www.opengis.net/def/crs/EPSG/0/4326",
  "extent": {
    "spatial": {
      "bbox": [[-180, -90, 180, 90]]
    }
  },
  "links": [
    {
      "rel": "self",
      "type": "application/json",
      "href": "https://maps.gnosis.earth/ogcapi/collections/blueMarble?f=json"
    },
    {
      "rel": "http://www.opengis.net/def/rel/ogc/1.0/coverage",
      "type": "image/tiff; application=geotiff",
      "href": "https://maps.gnosis.earth/ogcapi/collections/blueMarble/coverage?f=tif"
    },
    {
      "rel": "http://www.opengis.net/def/rel/ogc/1.0/coverage-rangetype",
      "type": "application/json",
      "href": "https://maps.gnosis.earth/ogcapi/collections/blueMarble/coverage/rangetype?f=json"
    }
  ]
}
``` 

QGIS 3.38.0
Windows 11

**Comment by doublebyte1** [comment]
- GDAL: issue with concatenation of OAPIF url
- QGIS: UI expects vector tiles on OGCAPI (it should use OAPIF, instead)
- QGIS: vector tiles gets stuck

**Comment by doublebyte1** [comment]
Status of the Add Layer (Raster and Vector) OGCAPI option using 3.39.0-Master [399f7df1c7](https://github.com/qgis/QGIS/commit/399f7df1c7) and [GDAL3.10.0dev-126a88523a](https://github.com/doublebyte1/gdal/commit/126a88523aec0b18540eb79ec887119224d356cc):

- OGC API - Maps ✅
- OGC API - Coverages ✅
- OGC API - Tiles (raster) ✅
- OGC API - Tiles (vector) ❌ - see [this](https://github.com/qgis/QGIS/issues/58174) issue
- OGC API Features ❌ - see [this](https://github.com/qgis/QGIS/issues/58087) issue

**Comment by doublebyte1** [comment]
Some the details to reproduce the tests above:

## OGC API - Maps

- Go to "Add Raster Layer..." and choose OGC API. On the endpoint, add https://maps.gnosis.earth/ogcapi/collections/blueMarble and on API options, choose `MAPS`.

![Screenshot from 2024-07-18 20-43-17](https://github.com/user-attachments/assets/ffb2c7c9-cd62-4ddf-ad25-763810ccea97)

## OGC API - Tiles (raster)

- Go to "Add Raster Layer..." and choose OGC API. On the endpoint, add https://maps.gnosis.earth/ogcapi/collections/blueMarble and on API options, choose `TILES`.

![Screenshot from 2024-07-18 20-43-52](https://github.com/user-attachments/assets/29464959-aebd-413b-82a4-110f2fed80a6)

## OGC API - Coverages

- Go to "Add Raster Layer..." and choose OGC API. On the endpoint, add https://maps.gnosis.earth/ogcapi/collections/HRDEM-Ottawa and on API options, choose `COVERAGES`.

![Screenshot from 2024-07-18 20-46-48](https://github.com/user-attachments/assets/435d10f4-3d8f-4585-8055-722c38a441cf)
 `COVERAGES`.

## OGC API - Features

More details on [this](https://github.com/qgis/QGIS/issues/58087) issue.

## OGC API - Tiles (vector)

More details on [this](https://github.com/qgis/QGIS/issues/58174) issue.

**Comment by kudlav** [comment]
Support of OGC API - Tiles (raster) is only partial as QGIS cannot load tileset from tileset description.

**Comment by jerstlouis** [comment]
@kudlav I don't think GDAL supports pointing it directly to a tileset end-point yet (e.g., https://maps.gnosis.earth/ogcapi/collections/Daraa/map/tiles/WebMercatorQuad). Perhaps we need to file an issue for GDAL about this first?

An implementation can actually conform to OGC API - Tiles "Core" without even providing tileset metadata.

If we really want to be comprehensive, it should also be possible to open a tileset by providing a template using variables for level, row, column (e.g., `https://maps.gnosis.earth/ogcapi/collections/Daraa/map/tiles/WebMercatorQuad/{tileMatrix}/{tileRow}/{tileCol}` plus the URI of a registered 2DTMS (e.g., https://www.opengis.net/def/tilematrixset/OGC/1.0/WebMercatorQuad) or a link or local 2DTMS 2.0 JSON definition (e.g., same link which can also return a JSON definition if following re-direction and using `Accept: application/json` request header, or somewhere else like https://raw.githubusercontent.com/opengeospatial/2D-Tile-Matrix-Set/master/registry/json/WebMercatorQuad.json).
